It happened crash after upgrade the version to 1.8.0

### Terraform Version

```shell
1.8.0
```


### Terraform Configuration Files

```terraform
...terraform config...
```


### Debug Output

```
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

panic: no expansion has been registered for module.cert_lb
goroutine 19550 [running]:
runtime/debug.Stack()
	/Users/runner/hostedtoolcache/go/1.22.1/x64/src/runtime/debug/stack.go:24 +0x5e
github.com/hashicorp/terraform/internal/logging.PanicHandler()
	/Users/runner/work/terraform/terraform/internal/logging/panic.go:84 +0x18b
panic({0x7213d00?, 0xc00592bef0?})
	/Users/runner/hostedtoolcache/go/1.22.1/x64/src/runtime/panic.go:770 +0x132
github.com/hashicorp/terraform/internal/instances.(*Expander).ExpandAbsModuleCall(0x734ed40?, {{0x0, 0x0, 0x0}, {{}, {0xc00013c8d0, 0x7}}})
	/Users/runner/work/terraform/terraform/internal/instances/expander.go:171 +0x265
github.com/hashicorp/terraform/internal/terraform.(*evaluationStateData).GetModule(0xc00159d560, {{}, {0xc00013c8d0?, 0x4a5f547?}}, {{0xc00072a970, 0xf}, {0x3b, 0x1e, 0x892}, {0x3b, ...}})
	/Users/runner/work/terraform/terraform/internal/terraform/evaluate.go:398 +0x788
github.com/hashicorp/terraform/internal/lang.(*Scope).evalContext(0xc00159d5f0, {0xc00aa8d320, 0x5, 0x5}, {0x0, 0x0})
	/Users/runner/work/terraform/terraform/internal/lang/eval.go:381 +0x1112
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalContext(...)
	/Users/runner/work/terraform/terraform/internal/lang/eval.go:245
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalBlock(0xc00159d5f0, {0x78fb640, 0xc002ac4730}, 0xc004d29980)
	/Users/runner/work/terraform/terraform/internal/lang/eval.go:58 +0xc5
github.com/hashicorp/terraform/internal/terraform.(*BuiltinEvalContext).EvaluateBlock(0x1?, {0x78fb8d0, 0xc00123edc0}, 0xc004d29980, {0x0?, 0x0?}, {{{{0x0, 0x0}}, {0x0, 0x0}}, ...})
	/Users/runner/work/terraform/terraform/internal/terraform/eval_context_builtin.go:315 +0x145
github.com/hashicorp/terraform/internal/terraform.(*NodeAbstractResourceInstance).plan(0xc005e24248, {0x791b128, 0xc000e2ec30}, 0x0, 0x0, 0x0, {0x0, 0x0, 0x0?})
	/Users/runner/work/terraform/terraform/internal/terraform/node_resource_abstract_instance.go:808 +0xb8b
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstance).managedResourceExecute(0xc005dd7340, {0x791b128, 0xc000e2ec30})
	/Users/runner/work/terraform/terraform/internal/terraform/node_resource_plan_instance.go:250 +0x1395
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstance).Execute(0x18?, {0x791b128?, 0xc000e2ec30?}, 0x0?)
	/Users/runner/work/terraform/terraform/internal/terraform/node_resource_plan_instance.go:75 +0x8a
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc003821cc0, {0x791b128, 0xc000e2ec30}, {0x554ec930, 0xc005dd7340})
	/Users/runner/work/terraform/terraform/internal/terraform/graph_walk_context.go:153 +0xbb
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x7880a60, 0xc005dd7340})
	/Users/runner/work/terraform/terraform/internal/terraform/graph.go:112 +0x62a
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc005d896e0, {0x7880a60, 0xc005dd7340}, 0xc00579f1c0)
	/Users/runner/work/terraform/terraform/internal/dag/walk.go:384 +0x2d7
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 15667
	/Users/runner/work/terraform/terraform/internal/dag/walk.go:307 +0xff3
```

### Expected Behavior

Nothing should have happened.

### Actual Behavior

It happened crash

### Steps to Reproduce

terraform plan

### Additional Context

_No response_

### References

_No response_

> Hi @georgepstaylor, thanks for the info! Do you have a reproducible config for the crash? I haven't been able to reproduce it yet so can't provide any guidance or potential workarounds at the moment. Thanks!

Hi @liamcervante.

https://github.com/ministryofjustice/modernisation-platform-environments/tree/main/terraform/environments/delius-core 
The project here causes a crash when applying.

```

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

panic: no expansion has been registered for module.environment_dev[0].module.oracle_db_primary
goroutine 6115 [running]:
runtime/debug.Stack()
	/opt/hostedtoolcache/go/1.22.1/x64/src/runtime/debug/stack.go:24 +0x5e
github.com/hashicorp/terraform/internal/logging.PanicHandler()
	/home/runner/work/terraform/terraform/internal/logging/panic.go:84 +0x18b
panic({0x2e00120?, 0xc00a84ce10?})
	/opt/hostedtoolcache/go/1.22.1/x64/src/runtime/panic.go:770 +0x132
github.com/hashicorp/terraform/internal/instances.(*Expander).ExpandAbsModuleCall(0x2f3a180?, {{0xc007b6c9c0, 0x1, 0x2}, {{}, {0xc0012e1f08, 0x11}}})
	/home/runner/work/terraform/terraform/internal/instances/expander.go:171 +0x265
github.com/hashicorp/terraform/internal/terraform.(*evaluationStateData).GetModule(0xc00d85e5a0, {{}, {0xc0012e1f08?, 0x7?}}, {{0xc00098cea0, 0x26}, {0x64, 0x5e, 0xc4d}, {0x64, ...}})
	/home/runner/work/terraform/terraform/internal/terraform/evaluate.go:398 +0x788
github.com/hashicorp/terraform/internal/lang.(*Scope).evalContext(0xc00d85e630, {0xc00d955488, 0x3, 0x3}, {0x0, 0x0})
	/home/runner/work/terraform/terraform/internal/lang/eval.go:381 +0x1112
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalContext(...)
	/home/runner/work/terraform/terraform/internal/lang/eval.go:245
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalExpr(0xc00d85e630, {0x3d75c00, 0xc000c12ae0}, {{0x3d75ce0?, 0x5891580?}})
	/home/runner/work/terraform/terraform/internal/lang/eval.go:170 +0x9e
github.com/hashicorp/terraform/internal/terraform.(*nodeModuleVariable).evalModuleVariable(0xc009daaee0, {0x3d984a8?, 0xc0006bdd10?}, 0xd8?)
	/home/runner/work/terraform/terraform/internal/terraform/node_module_variable.go:275 +0x1a5
github.com/hashicorp/terraform/internal/terraform.(*nodeModuleVariable).Execute(0xc009daaee0, {0x3d984a8, 0xc0006bdd10}, 0x1)
	/home/runner/work/terraform/terraform/internal/terraform/node_module_variable.go:214 +0x1c8
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc004d688c0, {0x3d984a8, 0xc0006bdd10}, {0x7f5f296fbd00, 0xc009daaee0})
	/home/runner/work/terraform/terraform/internal/terraform/graph_walk_context.go:153 +0xbb
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x31d97a0, 0xc009daaee0})
	/home/runner/work/terraform/terraform/internal/terraform/graph.go:112 +0x62a
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc000f66780, {0x31d97a0, 0xc009daaee0}, 0xc009df00c0)
	/home/runner/work/terraform/terraform/internal/dag/walk.go:384 +0x2d7
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 3929
	/home/runner/work/terraform/terraform/internal/dag/walk.go:307 +0xff3
```

The module source can be found here: https://github.com/ministryofjustice/modernisation-platform-environments/tree/main/terraform/environments/delius-core/modules/components/oracle_db_instance

Not sure if that's helpful?

--- a/internal/terraform/context_apply2_test.go
+++ b/internal/terraform/context_apply2_test.go
@@ -2893,3 +2893,35 @@ resource "test_object" "a" {
 		t.Errorf("Unexpected %s change for %s", c.Action, c.Addr)
 	}
 }
+
+// This test explicitly reproduces the issue described in #34976.
+func TestContext2Apply_34976(t *testing.T) {
+	m := testModuleInline(t, map[string]string{
+		"main.tf": `
+module "a" {
+  source = "./mod"
+  count = 1
+}
+
+resource "test_object" "obj" {
+  test_number = length(module.a)
+}
+`,
+		"mod/main.tf": ``, // just an empty module
+	})
+
+	p := simpleMockProvider()
+
+	ctx := testContext2(t, &ContextOpts{
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("test"): testProviderFuncFixed(p),
+		},
+	})
+
+	plan, diags := ctx.Plan(m, states.NewState(), SimplePlanOpts(plans.NormalMode, testInputValuesUnset(m.Module.Variables)))
+	assertNoErrors(t, diags)
+
+	// Just don't crash.
+	_, diags = ctx.Apply(plan, m, nil)
+	assertNoErrors(t, diags)
+}

--- a/internal/terraform/context_plan2_test.go
+++ b/internal/terraform/context_plan2_test.go
@@ -10,6 +10,7 @@ import (
 	"strings"
 	"sync"
 	"testing"
+	"time"
 
 	"github.com/davecgh/go-spew/spew"
 	"github.com/google/go-cmp/cmp"
@@ -5493,4 +5494,50 @@ resource "test_object" "obj" {
 	if ch.Action != plans.NoOp {
 		t.Fatal("expected no change in plan")
 	}
+}
+
+// This test explicitly reproduces the issue described in #34976.
+func TestContext2Plan_34976(t *testing.T) {
+	m := testModuleInline(t, map[string]string{
+		"main.tf": `
+data "test_object" "obj" {}
+
+module "a" {
+  depends_on = [data.test_object.obj]
+  source = "./mod"
+}
+
+output "value" {
+  value = try(module.a.notreal, null)
+}
+`,
+		"mod/main.tf": ``,
+	})
+
+	p := new(testing_provider.MockProvider)
+	p.GetProviderSchemaResponse = getProviderSchemaResponseFromProviderSchema(&providerSchema{
+		DataSources: map[string]*configschema.Block{
+			"test_object": {},
+		},
+	})
+	p.ReadDataSourceFn = func(req providers.ReadDataSourceRequest) providers.ReadDataSourceResponse {
+
+		// Just very small delay that means the output will be processed before
+		// the module.
+		time.Sleep(50 * time.Millisecond)
+
+		return providers.ReadDataSourceResponse{
+			State: req.Config,
+		}
+	}
+
+	ctx := testContext2(t, &ContextOpts{
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("test"): testProviderFuncFixed(p),
+		},
+	})
+
+	// Just shouldn't crash.
+	_, diags := ctx.Plan(m, states.NewState(), SimplePlanOpts(plans.NormalMode, testInputValuesUnset(m.Module.Variables)))
+	assertNoErrors(t, diags)
 }

--- a/internal/terraform/node_module_expand.go
+++ b/internal/terraform/node_module_expand.go
@@ -26,6 +26,7 @@ type nodeExpandModule struct {
 
 var (
 	_ GraphNodeExecutable       = (*nodeExpandModule)(nil)
+	_ GraphNodeReferenceable    = (*nodeExpandModule)(nil)
 	_ GraphNodeReferencer       = (*nodeExpandModule)(nil)
 	_ GraphNodeReferenceOutside = (*nodeExpandModule)(nil)
 	_ graphNodeExpandsInstances = (*nodeExpandModule)(nil)
@@ -74,6 +75,14 @@ func (n *nodeExpandModule) References() []*addrs.Reference {
 	return refs
 }
 
+func (n *nodeExpandModule) ReferenceableAddrs() []addrs.Referenceable {
+	// Anything referencing this module must do so after the ExpandModule call
+	// has been made to the expander, so we return the module call address as
+	// the only referenceable address.
+	_, call := n.Addr.Call()
+	return []addrs.Referenceable{call}
+}
+
 func (n *nodeExpandModule) DependsOn() []*addrs.Reference {
 	if n.ModuleCall == nil {
 		return nil
@@ -98,7 +107,7 @@ func (n *nodeExpandModule) DependsOn() []*addrs.Reference {
 
 // GraphNodeReferenceOutside
 func (n *nodeExpandModule) ReferenceOutside() (selfPath, referencePath addrs.Module) {
-	return n.Addr, n.Addr.Parent()
+	return n.Addr.Parent(), n.Addr.Parent()
 }
 
 // GraphNodeExecutable

--- a/internal/terraform/transform_reference.go
+++ b/internal/terraform/transform_reference.go
@@ -522,21 +522,50 @@ func (m ReferenceMap) referenceMapKey(path addrs.Module, addr addrs.Referenceabl
 		// might be in a resource-oriented graph rather than an
 		// instance-oriented graph, and so we'll see if we have the
 		// resource itself instead.
-		switch ri := addr.(type) {
-		case addrs.ResourceInstance:
-			addr = ri.ContainingResource()
-		case addrs.ResourceInstancePhase:
-			addr = ri.ContainingResource()
-		case addrs.ModuleCallInstanceOutput:
-			addr = ri.ModuleCallOutput()
-		case addrs.ModuleCallInstance:
-			addr = ri.Call
-		default:
-			return key
+
+		if ri, ok := addr.(addrs.ResourceInstance); ok {
+			return m.mapKey(path, ri.ContainingResource())
 		}
-		// if we matched any of the resource node types above, generate a new
-		// key
-		key = m.mapKey(path, addr)
+
+		if rip, ok := addr.(addrs.ResourceInstancePhase); ok {
+			return m.mapKey(path, rip.ContainingResource())
+		}
+
+		if mcio, ok := addr.(addrs.ModuleCallInstanceOutput); ok {
+
+			// A module call instance output is a reference to an output of a
+			// specific module call. If we can't find that, we'll look first
+			// for the general non-instanced output.
+
+			key = m.mapKey(path, mcio.ModuleCallOutput())
+			if _, exists := m[key]; exists {
+				// We found it, so we can just use that.
+				return key
+			}
+
+			// Otherwise we'll look just for the instanced module call itself.
+
+			key = m.mapKey(path, mcio.Call)
+			if _, exists := m[key]; exists {
+				// We found it, so we can just use that.
+				return key
+			}
+
+			// If we still can't find it, then we'll look for the non-instanced
+			// module call. This is the same as we'd do if the original call had
+			// just been for a ModuleCallInstance, so we'll let that fall
+			// through.
+
+			addr = mcio.Call
+
+		}
+
+		if mci, ok := addr.(addrs.ModuleCallInstance); ok {
+			return m.mapKey(path, mci.Call)
+		}
+
+		// If nothing matched, then we'll just return the original key
+		// unchanged.
 	}
 	return key
 }

